rfcs/2023-rsp-testing-utils.md

```diff
@@ -0,0 +1,113 @@
+
+In general, there are several internal test utils [here](https://github.com/adobe/react-spectrum/blob/07f673acb2d68144156df3aa0db6e91810b12c67/packages/dev/test-utils/src/events.ts) and in our tests that we should consider exposing that handle issuing/simulating the press/touch events that would happen in the browser. Ideally, using the `user-event` library would eliminate the need for most, if not all, of the interaction related util functions, but has proved insufficient in various cases (like for long press). We'll have to convert our own tests to use the `user-event` v14 and see if any other internal test utils can be replaced entirely.
+
+These util functions would live in one of two packages perhaps, `@react-aria/test-utils` and `@react-spectrum/test-utils`. The differentiating factor would be if the util is specific to our React Spectrum components (`simulateMobile`/`simulateDesktop`) or not (`triggerLongPress`).
+
+Caveat: The high level test util classes that are discussed below may also minimize the need for some of these interaction utils, open for discussion.
+
+
+3. Create high level test util classes to handle common interaction patterns
+
+Many of our components have several core user interaction patterns. For a ComboBox or a Picker, a user will usually open the dropdown via a click, look for a option in the list, and select that option. This is a pretty simple interaction but requires a fair bit of prerequisite knowledge to properly execute in a test. The test writer has to
+find the right element to press, wait for the dropdown to open and transition in, find the listbox and desired option, and finally click on said option and wait for the dropdown to transition out. Sometimes using `user-event` backfires and the order/timing of the underlying events fired doesn't match browser behavior, resulting in a test-env only failing test that is extremely hard to debug.
+
+To solve this, we could create a test helper class for each ARIA pattern (i.e. table, list, select, button, etc) that contains methods for common interactions specific to the pattern as described in [this issue comment](https://github.com/adobe/react-spectrum/issues/3703#issuecomment-1302727331). The interaction methods would need to work for `realTimers` and `fakeTimers` (if using Jest) and would handle all the requisite element lookup and event firing required for the high level interaction pattern. Where ever possible, these interaction methods should use `user-event` or our other interaction utils. Additionally, it may be valuable for each of the classes to look up and expose the internal nodes that users often want to assert against (e.g. The picker's trigger button, label, listbox, etc). A implementation of this could look like the following:
```

Comment:
Supporting both real and fake timers might be a lot of work. We could consider only supporting fake timers, at least to start. Another alternative could potentially be to turn on fake timers just before performing an interaction, and restore it afterward if a larger test is using real timers. That way we could potentially avoid having two paths.

The downside to that might be that it's specific to Jest, and if someone is using a test framework other than Jest it wouldn't work. The more we can reduce the complexity and number of scenarios we need to cover though the better.

Reply:
I'll look into this, definitely would be nice to reduce the complexity for a first iteration.

>  Another alternative could potentially be to turn on fake timers just before performing an interaction, and restore it afterward if a larger test is using real timers

An interesting idea, I'll have to do some testing around this. Not sure the ramifications of performing timer type switches since I've primarily had experience with sticking with fakeTimers, will have to play around with that.

> The downside to that might be that it's specific to Jest, and if someone is using a test framework other than Jest it wouldn't work. The more we can reduce the complexity and number of scenarios we need to cover though the better.

Yeah, that is a big concern of mine. Will need to test against some other test frameworks, will be a big research task.